File: tools/release/optimize_native_runtime_payload.py
#!/usr/bin/env python3
"""Prune, strip, and validate liboliphaunt native runtime payloads."""

from __future__ import annotations

import argparse
import os
import re
import shutil
import subprocess
import sys
from dataclasses import dataclass
from pathlib import Path, PurePosixPath
from typing import Literal, NoReturn

import strip_native_release_binaries


ROOT = Path(__file__).resolve().parents[2]
NATIVE_RUNTIME_TOOL_STEMS = ("initdb", "pg_ctl", "postgres")
NATIVE_TOOLS_TOOL_STEMS = ("pg_dump", "psql")
NATIVE_PACKAGED_TOOL_STEMS = (*NATIVE_RUNTIME_TOOL_STEMS, *NATIVE_TOOLS_TOOL_STEMS)
NativeToolSet = Literal["packaged", "runtime", "tools"]
ELF_DEBUG_SECTION = re.compile(r"\]\s+\.(debug_[^\s]+|symtab|strtab)\s")
MACHO_MAGICS = {
    b"\xfe\xed\xfa\xce",
    b"\xce\xfa\xed\xfe",
    b"\xfe\xed\xfa\xcf",
    b"\xcf\xfa\xed\xfe",
    b"\xca\xfe\xba\xbe",
    b"\xbe\xba\xfe\xca",
}
DEV_RUNTIME_DIRS = (
    PurePosixPath("include"),
    PurePosixPath("lib/pkgconfig"),
    PurePosixPath("lib/postgresql/pgxs"),
)
DEV_RUNTIME_SUFFIXES = (".a", ".la", ".pdb")
WINDOWS_DEV_RUNTIME_SUFFIXES = (".lib",)


@dataclass(frozen=True)
class NativeFile:
    path: Path
    kind: str
    archive: bool = False


def fail(message: str) -> NoReturn:
    print(f"optimize_native_runtime_payload.py: {message}", file=sys.stderr)
    raise SystemExit(1)


def rel(path: Path) -> str:
    try:
        return path.relative_to(ROOT).as_posix()
    except ValueError:
        return str(path)


def read_prefix(path: Path, size: int = 8) -> bytes:
    try:
        with path.open("rb") as file:
            return file.read(size)
    except OSError as error:
        fail(f"failed to read {path}: {error}")


def classify_native_file(path: Path) -> NativeFile | None:
    prefix = read_prefix(path)
    if prefix.startswith(b"\x7fELF"):
        return NativeFile(path, "elf")
    if prefix[:4] in MACHO_MAGICS:
        return NativeFile(path, "macho")
    if prefix.startswith(b"MZ"):
        return NativeFile(path, "pe")
    if prefix.startswith(b"!<arch>\n"):
        return NativeFile(path, "archive", archive=True)
    return None


def is_windows_target(target: str | None, runtime_dir: Path | None = None) -> bool:
    if target is not None and target.startswith("windows-"):
        return True
    if runtime_dir is None:
        return False
    bin_dir = runtime_dir / "bin"
    return any((bin_dir / f"{stem}.exe").exists() for stem in NATIVE_PACKAGED_TOOL_STEMS)


def required_runtime_tools(target: str | None, runtime_dir: Path | None = None) -> tuple[str, ...]:
    if is_windows_target(target, runtime_dir):
        return tuple(f"{stem}.exe" for stem in NATIVE_RUNTIME_TOOL_STEMS)
    return NATIVE_RUNTIME_TOOL_STEMS


def required_tools_package_tools(
    target: str | None, runtime_dir: Path | None = None
) -> tuple[str, ...]:
    if is_windows_target(target, runtime_dir):
        return tuple(f"{stem}.exe" for stem in NATIVE_TOOLS_TOOL_STEMS)
    return NATIVE_TOOLS_TOOL_STEMS


def packaged_runtime_tools(target: str | None, runtime_dir: Path | None = None) -> tuple[str, ...]:
    if is_windows_target(target, runtime_dir):
        return tuple(f"{stem}.exe" for stem in NATIVE_PACKAGED_TOOL_STEMS)
    return NATIVE_PACKAGED_TOOL_STEMS


def runtime_tools_for_set(
    target: str | None,
    runtime_dir: Path | None = None,
    *,
    tool_set: NativeToolSet = "packaged",
) -> tuple[str, ...]:
    if tool_set == "runtime":
        return required_runtime_tools(target, runtime_dir)
    if tool_set == "tools":
        return required_tools_package_tools(target, runtime_dir)
    return packaged_runtime_tools(target, runtime_dir)


def required_runtime_member_paths(target: str | None, *, prefix: str) -> list[str]:
    return [f"{prefix.rstrip('/')}/{tool}" for tool in required_runtime_tools(target)]


def required_tools_member_paths(target: str | None, *, prefix: str) -> list[str]:
    return [f"{prefix.rstrip('/')}/{tool}" for tool in required_tools_package_tools(target)]


def runtime_dir_for(root: Path) -> Path | None:
    for candidate in [
        root / "runtime",
        root / "oliphaunt" / "runtime" / "files",
    ]:
        if candidate.is_dir():
            return candidate
    if (root / "bin").is_dir() and ((root / "share").is_dir() or (root / "lib").is_dir()):
        return root
    return None


def remove_path(path: Path) -> None:
    if path.is_dir():
        shutil.rmtree(path)
    elif path.exists():
        path.unlink()


def prune_empty_dirs(root: Path) -> None:
    if not root.is_dir():
        return
    for path in sorted((item for item in root.rglob("*") if item.is_dir()), reverse=True):
        try:
            path.rmdir()
        except OSError:
            pass


def is_dev_runtime_file(relative: PurePosixPath, *, windows: bool) -> bool:
    name = relative.name.lower()
    if name.endswith(DEV_RUNTIME_SUFFIXES):
        return True
    if windows and name.endswith(WINDOWS_DEV_RUNTIME_SUFFIXES):
        return True
    return False


def prune_runtime_payload(
    root: Path,
    target: str | None = None,
    *,
    tool_set: NativeToolSet = "packaged",
) -> None:
    runtime_dir = runtime_dir_for(root)
    if runtime_dir is None:
        return

    windows = is_windows_target(target, runtime_dir)
    required_tools = set(runtime_tools_for_set(target, runtime_dir, tool_set=tool_set))
    bin_dir = runtime_dir / "bin"
    if bin_dir.is_dir():
        for path in sorted(bin_dir.iterdir()):
            name = path.name
            if windows:
                if name.lower().endswith(".exe") and name not in required_tools:
                    remove_path(path)
            elif name not in required_tools:
                remove_path(path)

    for relative in DEV_RUNTIME_DIRS:
        remove_path(runtime_dir.joinpath(*relative.parts))

    for path in sorted(runtime_dir.rglob("*"), reverse=True):
        if path.is_dir() and path.name.endswith(".dSYM"):
            remove_path(path)
            continue
        if not path.is_file():
            continue
        relative = PurePosixPath(path.relative_to(runtime_dir).as_posix())
        if is_dev_runtime_file(relative, windows=windows):
            remove_path(path)

    prune_empty_dirs(runtime_dir)


def strip_supported_for_target(target: str | None) -> bool:
    if target is None:
        return True
    if target.startswith(("linux-", "android-")):
        return sys.platform.startswith("linux")
    if target.startswith(("macos-", "ios-")):
        return sys.platform == "darwin"
    if target.startswith("windows-"):
        return bool(
            os.environ.get("OLIPHAUNT_PE_STRIP")
            or os.environ.get("OLIPHAUNT_STRIP")
            or shutil.which("llvm-strip")
            or sys.platform == "win32"
        )
    return True


def strip_payload(root: Path) -> None:
    result = strip_native_release_binaries.main([str(root)])
    if result != 0:
        fail(f"failed to strip native payload under {rel(root)}")


def iter_files(root: Path) -> list[Path]:
    return sorted(path for path in root.rglob("*") if path.is_file())


def file_output(path: Path) -> str | None:
    file_tool = shutil.which("file")
    if file_tool is None:
        return None
    result = subprocess.run(
        [file_tool, str(path)],
        check=False,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )
    if result.returncode != 0:
        return None
    return result.stdout


def elf_debug_errors(path: Path) -> list[str]:
    readelf = shutil.which("readelf")
    if readelf is not None:
        result = subprocess.run(
            [readelf, "-S", str(path)],
            check=False,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )
        if result.returncode != 0:
            return [f"{rel(path)} could not be inspected with readelf: {result.stderr.strip()}"]
        sections = sorted({match.group(1) for match in ELF_DEBUG_SECTION.finditer(result.stdout)})
        return [f"{rel(path)} contains unstripped ELF section .{section}" for section in sections]

    output = file_output(path)
    if output is not None and ("not stripped" in output or "with debug_info" in output):
        return [f"{rel(path)} appears to contain unstripped ELF debug/symbol data"]
    return []


def validate_native_files(root: Path) -> list[str]:
    errors: list[str] = []
    for path in iter_files(root):
        native = classify_native_file(path)
        if native is None:
            continue
        if native.kind == "elf" and not native.archive:
            errors.extend(elf_debug_errors(path))
    return errors


def validate_runtime_tree(
    root: Path,
    target: str | None,
    require_runtime: bool,
    *,
    tool_set: NativeToolSet = "packaged",
) -> list[str]:
    errors: list[str] = []
    runtime_dir = runtime_dir_for(root)
    if runtime_dir is None:
        if require_runtime:
            errors.append(f"{rel(root)} is missing a runtime tree")
        return errors

    windows = is_windows_target(target, runtime_dir)
    required_tools = set(runtime_tools_for_set(target, runtime_dir, tool_set=tool_set))
    bin_dir = runtime_dir / "bin"
    if require_runtime and not bin_dir.is_dir():
        errors.append(f"{rel(runtime_dir)} is missing bin")
    if bin_dir.is_dir():
        for tool in sorted(required_tools):
            path = bin_dir / tool
            if not path.is_file():
                errors.append(f"{rel(runtime_dir)} is missing required runtime tool bin/{tool}")
                continue
            if not windows and not os.access(path, os.X_OK):
                errors.append(f"{rel(path)} must be executable")
        for path in sorted(bin_dir.iterdir()):
            if windows:
                if path.name.lower().endswith(".exe") and path.name not in required_tools:
                    errors.append(f"{rel(path)} is an extra Windows runtime executable")
            elif path.name not in required_tools:
                errors.append(f"{rel(path)} is an extra runtime tool")

    for relative in DEV_RUNTIME_DIRS:
        path = runtime_dir.joinpath(*relative.parts)
        if path.exists():
            errors.append(f"{rel(path)} is a development-only runtime path")

    for path in sorted(runtime_dir.rglob("*")):
        if path.is_dir() and path.name.endswith(".dSYM"):
            errors.append(f"{rel(path)} is a development-only debug symbol bundle")
            continue
        if not path.is_file():
            continue
        relative = PurePosixPath(path.relative_to(runtime_dir).as_posix())
        if is_dev_runtime_file(relative, windows=windows):
            errors.append(f"{rel(path)} is a development-only runtime file")

    return errors


def validate_payload(
    root: Path,
    target: str | None = None,
    *,
    require_runtime: bool = True,
    tool_set: NativeToolSet = "packaged",
) -> None:
    errors = [
        *validate_runtime_tree(
            root,
            target,
            require_runtime=require_runtime,
            tool_set=tool_set,
        ),
        *validate_native_files(root),
    ]
    if errors:
        for error in errors:
            print(error, file=sys.stderr)
        fail(f"{rel(root)} is not an optimized native runtime payload")


def optimize_payload(
    root: Path,
    target: str | None = None,
    *,
    strip: bool | Literal["auto"] = "auto",
    require_runtime: bool = True,
    tool_set: NativeToolSet = "packaged",
) -> None:
    prune_runtime_payload(root, target, tool_set=tool_set)
    should_strip = strip is True or (strip == "auto" and strip_supported_for_target(target))
    if should_strip:
        strip_payload(root)
    validate_payload(root, target, require_runtime=require_runtime, tool_set=tool_set)


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("root", type=Path)
    parser.add_argument("--target", default=None)
    parser.add_argument("--check", action="store_true", help="validate without mutating the payload")
    parser.add_argument(
        "--no-strip",
        action="store_true",
        help="prune but skip native binary stripping before validation",
    )
    parser.add_argument(
        "--allow-missing-runtime",
        action="store_true",
        help="validate native files even when the archive is a library-only mobile payload",
    )
    parser.add_argument(
        "--tool-set",
        choices=("packaged", "runtime", "tools"),
        default="packaged",
        help="which packaged runtime bin tools are expected in the payload",
    )
    return parser.parse_args(argv)


def main(argv: list[str]) -> int:
    args = parse_args(argv)
    root = args.root.resolve()
    if not root.exists():
        fail(f"payload root does not exist: {root}")
    if args.check:
        validate_payload(
            root,
            args.target,
            require_runtime=not args.allow_missing_runtime,
            tool_set=args.tool_set,
        )
        return 0
    optimize_payload(
        root,
        args.target,
        strip=False if args.no_strip else "auto",
        require_runtime=not args.allow_missing_runtime,
        tool_set=args.tool_set,
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: tools/release/optimize_native_runtime_payload.mjs
#!/usr/bin/env bun
import {
  accessSync,
  closeSync,
  constants,
  existsSync,
  lstatSync,
  openSync,
  readFileSync,
  readdirSync,
  readSync,
  rmSync,
  rmdirSync,
} from "node:fs";
import { dirname, join, relative, resolve, sep } from "node:path";
import { spawnSync } from "node:child_process";
import { platform } from "node:os";
import { fileURLToPath } from "node:url";

const TOOL = "optimize_native_runtime_payload.mjs";
const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "../..");
const POLICY_PATH = join(ROOT, "tools/release/native-runtime-payload-policy.json");
const POLICY = JSON.parse(readFileSync(POLICY_PATH, "utf8"));

export const NATIVE_RUNTIME_TOOL_STEMS = Object.freeze([...POLICY.nativeRuntimeToolStems]);
export const NATIVE_TOOLS_TOOL_STEMS = Object.freeze([...POLICY.nativeToolsToolStems]);
export const NATIVE_PACKAGED_TOOL_STEMS = Object.freeze([
  ...NATIVE_RUNTIME_TOOL_STEMS,
  ...NATIVE_TOOLS_TOOL_STEMS,
]);

const DEV_RUNTIME_DIRS = Object.freeze([...POLICY.devRuntimeDirs]);
const DEV_RUNTIME_SUFFIXES = Object.freeze([...POLICY.devRuntimeSuffixes]);
const WINDOWS_DEV_RUNTIME_SUFFIXES = Object.freeze([...POLICY.windowsDevRuntimeSuffixes]);
const MACHO_MAGICS = new Set([
  "feedface",
  "cefaedfe",
  "feedfacf",
  "cffaedfe",
  "cafebabe",
  "bebafeca",
]);
const ELF_DEBUG_SECTION = /\]\s+\.(debug_[^\s]+|symtab|strtab)\s/g;

function fail(message) {
  console.error(`${TOOL}: ${message}`);
  process.exit(1);
}

function rel(path) {
  const resolved = resolve(String(path));
  const relativePath = relative(ROOT, resolved);
  if (!relativePath || relativePath.startsWith("..") || relativePath === resolved) {
    return resolved.split(sep).join("/");
  }
  return relativePath.split(sep).join("/");
}

function exists(path) {
  return existsSync(path);
}

function isDirectory(path) {
  try {
    return lstatSync(path).isDirectory();
  } catch {
    return false;
  }
}

function isFile(path) {
  try {
    return lstatSync(path).isFile();
  } catch {
    return false;
  }
}

function readPrefix(path, size = 8) {
  const buffer = Buffer.alloc(size);
  let fd;
  try {
    fd = openSync(path, "r");
    const bytesRead = readSync(fd, buffer, 0, size, 0);
    return buffer.subarray(0, bytesRead);
  } catch (error) {
    fail(`failed to read ${path}: ${error.message}`);
  } finally {
    if (fd !== undefined) {
      closeSync(fd);
    }
  }
}

function classifyNativeFile(path) {
  const prefix = readPrefix(path);
  if (prefix.subarray(0, 4).equals(Buffer.from([0x7f, 0x45, 0x4c, 0x46]))) {
    return { path, kind: "elf", archive: false };
  }
  if (MACHO_MAGICS.has(prefix.subarray(0, 4).toString("hex"))) {
    return { path, kind: "macho", archive: false };
  }
  if (prefix.subarray(0, 2).toString("ascii") === "MZ") {
    return { path, kind: "pe", archive: false };
  }
  if (prefix.subarray(0, 8).toString("ascii") === "!<arch>\n") {
    return { path, kind: "archive", archive: true };
  }
  return null;
}

export function isWindowsTarget(target, runtimeDir = null) {
  if (target && target.startsWith("windows-")) {
    return true;
  }
  if (!runtimeDir) {
    return false;
  }
  const binDir = join(runtimeDir, "bin");
  return NATIVE_PACKAGED_TOOL_STEMS.some((stem) => isFile(join(binDir, `${stem}.exe`)));
}

export function requiredRuntimeTools(target, runtimeDir = null) {
  if (isWindowsTarget(target, runtimeDir)) {
    return NATIVE_RUNTIME_TOOL_STEMS.map((stem) => `${stem}.exe`);
  }
  return [...NATIVE_RUNTIME_TOOL_STEMS];
}

export function requiredToolsPackageTools(target, runtimeDir = null) {
  if (isWindowsTarget(target, runtimeDir)) {
    return NATIVE_TOOLS_TOOL_STEMS.map((stem) => `${stem}.exe`);
  }
  return [...NATIVE_TOOLS_TOOL_STEMS];
}

export function packagedRuntimeTools(target, runtimeDir = null) {
  if (isWindowsTarget(target, runtimeDir)) {
    return NATIVE_PACKAGED_TOOL_STEMS.map((stem) => `${stem}.exe`);
  }
  return [...NATIVE_PACKAGED_TOOL_STEMS];
}

export function runtimeToolsForSet(target, runtimeDir = null, toolSet = "packaged") {
  if (toolSet === "runtime") {
    return requiredRuntimeTools(target, runtimeDir);
  }
  if (toolSet === "tools") {
    return requiredToolsPackageTools(target, runtimeDir);
  }
  return packagedRuntimeTools(target, runtimeDir);
}

export function requiredRuntimeMemberPaths(target, prefix) {
  return requiredRuntimeTools(target).map((tool) => `${prefix.replace(/\/+$/, "")}/${tool}`);
}

export function requiredToolsMemberPaths(target, prefix) {
  return requiredToolsPackageTools(target).map((tool) => `${prefix.replace(/\/+$/, "")}/${tool}`);
}

function runtimeDirFor(root) {
  for (const candidate of [
    join(root, "runtime"),
    join(root, "oliphaunt", "runtime", "files"),
  ]) {
    if (isDirectory(candidate)) {
      return candidate;
    }
  }
  if (isDirectory(join(root, "bin")) && (isDirectory(join(root, "share")) || isDirectory(join(root, "lib")))) {
    return root;
  }
  return null;
}

function removePath(path) {
  rmSync(path, { recursive: true, force: true });
}

function walk(root, { includeDirs = false } = {}) {
  if (!isDirectory(root)) {
    return [];
  }
  const results = [];
  const visit = (current) => {
    for (const name of readdirSync(current).sort()) {
      const path = join(current, name);
      let stat;
      try {
        stat = lstatSync(path);
      } catch {
        continue;
      }
      if (stat.isDirectory()) {
        if (includeDirs) {
          results.push(path);
        }
        visit(path);
      } else if (stat.isFile()) {
        results.push(path);
      }
    }
  };
  visit(root);
  return results.sort();
}

function pruneEmptyDirs(root) {
  for (const path of walk(root, { includeDirs: true }).filter(isDirectory).sort().reverse()) {
    try {
      rmdirSync(path);
    } catch {
      // Directory is not empty or disappeared while pruning.
    }
  }
}

function posixRelative(from, to) {
  return relative(from, to).split(sep).join("/");
}

function isDevRuntimeFile(relativePath, { windows }) {
  const name = relativePath.split("/").pop().toLowerCase();
  if (DEV_RUNTIME_SUFFIXES.some((suffix) => name.endsWith(suffix))) {
    return true;
  }
  return windows && WINDOWS_DEV_RUNTIME_SUFFIXES.some((suffix) => name.endsWith(suffix));
}

export function pruneRuntimePayload(root, target = null, { toolSet = "packaged" } = {}) {
  const runtimeDir = runtimeDirFor(root);
  if (!runtimeDir) {
    return;
  }

  const windows = isWindowsTarget(target, runtimeDir);
  const requiredTools = new Set(runtimeToolsForSet(target, runtimeDir, toolSet));
  const binDir = join(runtimeDir, "bin");
  if (isDirectory(binDir)) {
    for (const name of readdirSync(binDir).sort()) {
      const path = join(binDir, name);
      if (windows) {
        if (name.toLowerCase().endsWith(".exe") && !requiredTools.has(name)) {
          removePath(path);
        }
      } else if (!requiredTools.has(name)) {
        removePath(path);
      }
    }
  }

  if (toolSet === "tools" && isDirectory(runtimeDir)) {
    for (const name of readdirSync(runtimeDir).sort()) {
      if (name !== "bin") {
        removePath(join(runtimeDir, name));
      }
    }
  }

  for (const relativePath of DEV_RUNTIME_DIRS) {
    removePath(join(runtimeDir, ...relativePath.split("/")));
  }

  for (const path of walk(runtimeDir, { includeDirs: true }).sort().reverse()) {
    if (isDirectory(path) && path.endsWith(".dSYM")) {
      removePath(path);
      continue;
    }
    if (!isFile(path)) {
      continue;
    }
    const relativePath = posixRelative(runtimeDir, path);
    if (isDevRuntimeFile(relativePath, { windows })) {
      removePath(path);
    }
  }

  pruneEmptyDirs(runtimeDir);
}

function which(command) {
  const pathEnv = process.env.PATH ?? "";
  const extensions = platform() === "win32" ? ["", ".exe", ".cmd", ".bat"] : [""];
  for (const dir of pathEnv.split(platform() === "win32" ? ";" : ":")) {
    if (!dir) {
      continue;
    }
    for (const extension of extensions) {
      const candidate = join(dir, `${command}${extension}`);
      if (isFile(candidate)) {
        return candidate;
      }
    }
  }
  return null;
}

function stripSupportedForTarget(target) {
  if (!target) {
    return true;
  }
  if (target.startsWith("linux-") || target.startsWith("android-")) {
    return platform() === "linux";
  }
  if (target.startsWith("macos-") || target.startsWith("ios-")) {
    return platform() === "darwin";
  }
  if (target.startsWith("windows-")) {
    return Boolean(
      process.env.OLIPHAUNT_PE_STRIP ||
        process.env.OLIPHAUNT_STRIP ||
        which("llvm-strip") ||
        platform() === "win32",
    );
  }
  return true;
}

function stripPayload(root) {
  const result = spawnSync(process.execPath, ["tools/release/strip_native_release_binaries.mjs", root], {
    cwd: ROOT,
    stdio: "inherit",
    env: process.env,
  });
  if (result.status !== 0) {
    fail(`failed to strip native payload under ${rel(root)}`);
  }
}

function fileOutput(path) {
  const fileTool = which("file");
  if (!fileTool) {
    return null;
  }
  const result = spawnSync(fileTool, [path], {
    cwd: ROOT,
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
  });
  if (result.status !== 0) {
    return null;
  }
  return result.stdout;
}

function elfDebugErrors(path) {
  const readelf = which("readelf");
  if (readelf) {
    const result = spawnSync(readelf, ["-S", path], {
      cwd: ROOT,
      encoding: "utf8",
      stdio: ["ignore", "pipe", "pipe"],
    });
    if (result.status !== 0) {
      return [`${rel(path)} could not be inspected with readelf: ${result.stderr.trim()}`];
    }
    const sections = new Set();
    for (const match of result.stdout.matchAll(ELF_DEBUG_SECTION)) {
      sections.add(match[1]);
    }
    return [...sections].sort().map((section) => `${rel(path)} contains unstripped ELF section .${section}`);
  }

  const output = fileOutput(path);
  if (output && (output.includes("not stripped") || output.includes("with debug_info"))) {
    return [`${rel(path)} appears to contain unstripped ELF debug/symbol data`];
  }
  return [];
}

function validateNativeFiles(root) {
  const errors = [];
  for (const path of walk(root)) {
    const native = classifyNativeFile(path);
    if (!native) {
      continue;
    }
    if (native.kind === "elf" && !native.archive) {
      errors.push(...elfDebugErrors(path));
    }
  }
  return errors;
}

function validateRuntimeTree(root, target, requireRuntime, { toolSet = "packaged" } = {}) {
  const errors = [];
  const runtimeDir = runtimeDirFor(root);
  if (!runtimeDir) {
    if (requireRuntime) {
      errors.push(`${rel(root)} is missing a runtime tree`);
    }
    return errors;
  }

  const windows = isWindowsTarget(target, runtimeDir);
  const requiredTools = new Set(runtimeToolsForSet(target, runtimeDir, toolSet));
  const binDir = join(runtimeDir, "bin");
  if (requireRuntime && !isDirectory(binDir)) {
    errors.push(`${rel(runtimeDir)} is missing bin`);
  }
  if (isDirectory(binDir)) {
    for (const tool of [...requiredTools].sort()) {
      const path = join(binDir, tool);
      if (!isFile(path)) {
        errors.push(`${rel(runtimeDir)} is missing required runtime tool bin/${tool}`);
        continue;
      }
      if (!windows) {
        try {
          accessSync(path, constants.X_OK);
        } catch {
          errors.push(`${rel(path)} must be executable`);
        }
      }
    }
    for (const name of readdirSync(binDir).sort()) {
      const path = join(binDir, name);
      if (windows) {
        if (name.toLowerCase().endsWith(".exe") && !requiredTools.has(name)) {
          errors.push(`${rel(path)} is an extra Windows runtime executable`);
        }
      } else if (!requiredTools.has(name)) {
        errors.push(`${rel(path)} is an extra runtime tool`);
      }
    }
  }

  if (toolSet === "tools" && isDirectory(runtimeDir)) {
    const allowed = new Set([...requiredTools].map((tool) => `bin/${tool}`));
    for (const path of walk(runtimeDir)) {
      const relativePath = posixRelative(runtimeDir, path);
      if (!allowed.has(relativePath)) {
        errors.push(`${rel(path)} is not part of the native tools payload`);
      }
    }
  }

  for (const relativePath of DEV_RUNTIME_DIRS) {
    const path = join(runtimeDir, ...relativePath.split("/"));
    if (exists(path)) {
      errors.push(`${rel(path)} is a development-only runtime path`);
    }
  }

  for (const path of walk(runtimeDir, { includeDirs: true })) {
    if (isDirectory(path) && path.endsWith(".dSYM")) {
      errors.push(`${rel(path)} is a development-only debug symbol bundle`);
      continue;
    }
    if (!isFile(path)) {
      continue;
    }
    const relativePath = posixRelative(runtimeDir, path);
    if (isDevRuntimeFile(relativePath, { windows })) {
      errors.push(`${rel(path)} is a development-only runtime file`);
    }
  }

  return errors;
}

export function validatePayload(root, target = null, { requireRuntime = true, toolSet = "packaged" } = {}) {
  const errors = [
    ...validateRuntimeTree(root, target, requireRuntime, { toolSet }),
    ...validateNativeFiles(root),
  ];
  if (errors.length > 0) {
    for (const error of errors) {
      console.error(error);
    }
    fail(`${rel(root)} is not an optimized native runtime payload`);
  }
}

export function optimizePayload(
  root,
  target = null,
  { strip = "auto", requireRuntime = true, toolSet = "packaged" } = {},
) {
  pruneRuntimePayload(root, target, { toolSet });
  const shouldStrip = strip === true || (strip === "auto" && stripSupportedForTarget(target));
  if (shouldStrip) {
    stripPayload(root);
  }
  validatePayload(root, target, { requireRuntime, toolSet });
}

function usage() {
  return `Usage: tools/release/optimize_native_runtime_payload.mjs <root> [options]

Prune, strip, and validate liboliphaunt native runtime payloads.

Options:
  --target <target>           Release target id.
  --check                     Validate without mutating the payload.
  --no-strip                  Prune but skip native binary stripping before validation.
  --allow-missing-runtime     Validate native files when the archive is library-only.
  --tool-set <set>            packaged, runtime, or tools. Default: packaged.
  --help                      Show this help.
`;
}

function parseArgs(argv) {
  const args = {
    root: null,
    target: null,
    check: false,
    noStrip: false,
    allowMissingRuntime: false,
    toolSet: "packaged",
  };
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === "--help" || arg === "-h") {
      console.log(usage());
      process.exit(0);
    }
    if (arg === "--target") {
      args.target = argv[++index];
      if (!args.target) {
        fail("--target requires a value");
      }
      continue;
    }
    if (arg === "--check") {
      args.check = true;
      continue;
    }
    if (arg === "--no-strip") {
      args.noStrip = true;
      continue;
    }
    if (arg === "--allow-missing-runtime") {
      args.allowMissingRuntime = true;
      continue;
    }
    if (arg === "--tool-set") {
      args.toolSet = argv[++index];
      if (!["packaged", "runtime", "tools"].includes(args.toolSet)) {
        fail("--tool-set must be one of: packaged, runtime, tools");
      }
      continue;
    }
    if (arg.startsWith("-")) {
      fail(`unknown option: ${arg}`);
    }
    if (args.root) {
      fail(`unexpected positional argument: ${arg}`);
    }
    args.root = arg;
  }
  if (!args.root) {
    console.error(usage());
    process.exit(2);
  }
  return args;
}

export function main(argv = process.argv.slice(2)) {
  const args = parseArgs(argv);
  const root = resolve(args.root);
  if (!exists(root)) {
    fail(`payload root does not exist: ${root}`);
  }
  if (args.check) {
    validatePayload(root, args.target, {
      requireRuntime: !args.allowMissingRuntime,
      toolSet: args.toolSet,
    });
    return;
  }
  optimizePayload(root, args.target, {
    strip: args.noStrip ? false : "auto",
    requireRuntime: !args.allowMissingRuntime,
    toolSet: args.toolSet,
  });
}

if (import.meta.main) {
  main();
}
